Adds behaviors to components

jacwright opened the pull request:
Behaviors add additional functionality to elements within your component's template that may be difficult to add with other mechanisms. Examples of functionality which behaviors makes trivial to attach are:
* tooltips
* image lazy loaders
* drag and drop functionality

Behaviors can be added to an element with the `add` directive.

```html
<img add:lazyload data-src="giant-photo.jpg>
```

Data may be passed to the behavior as an object literal (e.g. `add:b="{ setting: true }"`, a literal value (e.g. `add:b="'a string'"`), or a value or function from your component's state (e.g. `add:b="foo"` or `add:b="foo()"`).

Behaviors are defined in a "behaviors" property on your component definition.

```html
<script>
  export default {
    behaviors: {
      b(node, data) {
        // do something
        return {
          update(data) {},
          destroy() {}
        }
      }
    }
  }
</script>
```

A behavior is a function which receives a reference to an element and optionally the data if it is added in the HTML. This function can then attach listeners or alter the element as needed. The behavior can optionally return an object with the methods `update(data)` and `destroy()`.

When data is added in the HTML and comes from state, the behavior's `update(data)` will be called if defined whenever the state is changed.

When the element is removed from the DOM `destroy()` will be called if provided, allowing for cleanup of event listeners, etc.

See https://github.com/sveltejs/svelte/issues/469 for discussion around this feature and more examples of how it could be used.

<!--
Thank you for creating a pull request. Before submitting, please note the following:

* If your pull request implements a new feature, please raise an issue to discuss it before sending code. In many cases features are absent for a reason.
* This message body should clearly illustrate what problems it solves. If there are related issues, remember to reference them.
* Ideally, include a test that fails without this PR but passes with it. PRs will only be merged once they pass CI. (Remember to `npm run lint`!)
-->

emilos commented on src/generators/nodes/Element.ts:
```diff
@@ -12,13 +12,15 @@ import Binding from './Binding';
+import Behavior from './Behavior';
 import Text from './Text';
 import * as namespaces from '../../utils/namespaces';
+import Behavior from './Behavior';
```
two same imports `import Behavior from './Behavior';` ^

Rich-Harris commented:
Damn, @jacwright! This looks awesome. The PR is great on an initial read — just one bit of feedback, which is that generally it's better to add a `runtime` test than a `js` test, because a) it guarantees the correctness of the generated code over time, and b) there are fewer tests to update whenever we make changes that require the files in `js/samples` to be updated.

Also, we could make the generated code a tiny bit simpler:

```diff
-link_behavior = link(a);
+link_behavior = link(a) || {};

-if (link_behavior && typeof link_behavior.destroy === 'function') {
-  link_behavior.destroy();
-}
+if (link_behavior.destroy) link_behavior.destroy();
```

Onto the bikeshedding: one problem with 'behavior' is [that's the terminology we use](https://svelte.technology/guide#behaviour) to describe all of a component's encapsulated logic — methods, transitions, etc. Also, it's spelt 'behaviour' 😀.

Re-reading #469, I like the 'trait' suggestion for its English-language connotations, but less so for its connotations in programming. I also like `as` but you're probably right, it doesn't work in all contexts. So just to be awkward, I'm going to make an entirely new suggestion:

```html
<img with:lazyload data-src='potato.jpg'>

<script>
  export default {
    augmentations: {
      lazyload(node, data) {...}
    }
  };
</script>
```

Any thoughts, anyone?

I also want to make sure we fully think through the lifecycle stuff, because it'd be difficult to make changes in future. Are there situations where you might want to run code before the element enters or exits the DOM? I like the simplicity of `update` and `destroy`, but I wonder if perhaps it should be something like `mount`, `update`, `beforeunmount` and `unmount` instead. (`beforemount` would be the code that runs when it's first created.)

I can think of two fairly contrived examples off the top of my head — if you had `<img with:lazyload src='whatever.jpg'>`, then you'd want to remove the `src` attribute before mounting so that you could control when the loading happened. (Obviously there is an easy workaround there, am just thinking aloud.) Or you might want to preserve information about the node's layout while it's still in the DOM, so that you can apply a transition effect to an incoming component after the outgoing component has been destroyed (a la [ramjet](https://www.rich-harris.co.uk/ramjet/) — excuse the currently-broken demo).

CC'ing @evs-chris — Chris, do you have any thoughts on the lifecycle stuff? Any lessons from Ractive you could share?

paulocoghi commented:
> Any thoughts, anyone?

What do you think of one of these: "action(s)",  "enhancement(s)", "addition(s)" or "extension(s)"?

jacwright commented:
I like `traits` and `with`. I say we go with that or follow an existing convention and do `behaviors` and `as`.

I also like the simplicity of what is there with just update and destroy.  We can add additional lifecycle events later if we need. The one thing that will break backwards compatibility if is when the initial method is run. The default being before mount makes sense to me (many traits won't care about the DOM). What if we start with these options on the returned object:
* `mount`
* `update`
* `unmount` (before leaving the dom)
* `destroy`

And we can add `afterDestroy` if it is needed.

jacwright commented:
A problem I just thought about is hydration. You can't add a trait to an element _before_ it is in the DOM  under all circumstances. If your page is server-rendered you only get the element after it is in the page. Should we account for that and just provide `update` and `destroy` and then maybe, in the future, `unmount` if a use-case turns up needing it?

tomcon commented:
Am not so keen on: 
1. traits (as per Rich language connotations)
2. augmentations  - too verbose
3. with (no clear purpose when you read the word + language connotations)

Looking at your opening statement @jacwright : 
"Behaviors add additional functionality to elements" 
http://www.thesaurus.com/browse/functionality?s=t
= performance, process, purpose, service, *use*, component
Therefore this feels good imo
```
<img use:lazyload data-src='potato.jpg'>

<script>
  export default {
    using: {
      lazyload(node, data) {...}
    }
   or 
    uses: {
      lazyload(node, data) {...}
    },
  };
</script>
```

jacwright commented:
I've made the changes request @Rich-Harris.
* renamed behaviors to traits
* renamed `add:xyz` to `with:xyz`
* removed most JS tests and added runtime tests instead
* added additional lifecycle methods, `mount` and `unmount` for cases that need them

Note: elements that hydrate will not benefit from pre-mounting logic. I'm not sure if addEventListener and other DOM methods are available on the server or whether allowing traits to run on the server will cause issues. Originally my thinking was they were only client-side.

jacwright commented:
@tomcon I like `use`. But we would still need a noun to reference them by in the docs or libraries. I think decorators and traits are fair game, even if they mean something else at the language level. As it stands now, this PR is using `traits` and `with`. I _would_ like to come to a consensus or have @Rich-Harris make a call on it. The more important thing to me is having the functionality this provides.

Rich-Harris commented:
@tomcon's `use` and @paulocoghi's `actions` are excellent suggestions. I'd vote for that combo:

```html
<img use:lazyload data-src='potato.jpg'>

<script>
  export default {
    actions: {
      lazyload(node, data) {...}
    }
  };
</script>
```

How does that sound for everyone?

> A problem I just thought about is hydration. You can't add a trait to an element before it is in the DOM under all circumstances.

Gah, you're absolutely right. Having a pre-mount hook is liable to cause all sorts of confusion. Better to not have it at all than to have subtle behaviour differences that people will struggle to debug. And since `unmount` *is* something that could be added later as a non-breaking change, I'm going to revert to type (@tomcon gets me) and suggest we file `unmount` under [YAGNI](https://martinfowler.com/bliki/Yagni.html) for now. So, back to `update` and `destroy`? Sorry for the back and forth on this!

paulocoghi commented:
> How does that sound for everyone?

Sounds great to me!

jacwright commented:
Using `actions` and `use`. Only supporting `update` and `destroy`. Rebased and squashed to a single commit to keep it clean. I believe this PR is ready except for the following:

Getting errors on validation tests where:
`Cannot read property 'line' of undefined` should be `Missing action 'whatever'`
`Cannot read property 'line' of undefined` should be `Actions can only be applied to DOM elements, not components`

I cannot reproduce locally, even after a fresh `npm install` and clean test directory and switching to node 6. It passes every time for me. Any ideas? I assume the error is from `warning.loc` being undefined, but without reproing I can't troubleshoot.

Rich-Harris commented:
Nice! I would guess the test failures are related to #1250 — `validator.warn` and `validator.error` now take a Node (or `{ start, end }` object) as the second argument, rather than a number representing `start`

Rich-Harris commented on src/validate/html/validateElement.ts:
```diff
@@ -217,6 +217,19 @@ export default function validateElement(
+			if (!validator.actions.has(attribute.name)) {
+				validator.error(
+					`Missing action '${attribute.name}'`,
+					attribute.start
```
this should now just be

```js
validator.error(`Missing action '${attribute.name}'`, attribute)
```

due to #1250

Rich-Harris commented on src/validate/js/propValidators/actions.ts:
```diff
@@ -0,0 +1,16 @@
+	if (prop.value.type !== 'ObjectExpression') {
+		validator.error(
+			`The 'actions' property must be an object literal`,
+			prop.start
```
see previous comment — `prop`, instead of `prop.start`

jacwright commented:
All done. Ready to merge.

PaulMaly commented:
Does it look like a decorator plugin in Ractive, right?

jacwright commented:
@PaulMaly exactly.

constgen commented:
Tooltips, image lazy loaders, drag and drop functionality - these cases can be handled by components composition or component wrapper, do they?
Example:

```
<Tooltip title=""><img src="potato.jpg" /></Tooltip>
<LazyImage src="potato.jpg"><img /></LazyImage>
```

Lets not extend the framework with yet another syntax

jacwright commented:
Your tooltip component will have to wrap your image with a span tag or
something, it can’t just add events to its children. And if you are adding
multiple actions to it you will have to wrap it multiple times.

Your LazyLoad image is now inextensible. What if you want to add a class?
Perhaps the author of LazyLoad thought of that and sets className onto the
<img>. But will the author consider everything? Perhaps if we get
{...state} attributes.

I totally get not wanting to extend the syntax. I tried doing these things
and in practice it was not easy or pretty. Actions provide a much cleaner
and easier way to accomplish a certain set of functionality that would be
much more difficult without it. It is a great tool to have at your
disposal, and one which you don’t need to use and won’t add any code to
your app if you choose not to use.

On Fri, Mar 23, 2018 at 7:42 AM Constantine Genchevsky <
notifications@github.com> wrote:

> Tooltips, image lazy loaders, drag and drop functionality - these cases
> can be handled by components composition or component wrapper, do they?
> Example:
>
> <Tooltip title=""><img src="potato.jpg" /></Tooltip>
> <LazyImage src="potato.jpg"><img /></LazyImage>
>
> Lets not extend the framework with yet another syntax
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sveltejs/svelte/pull/1247#issuecomment-375668348>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHV5n3Yyt6VOLGDMy2NSJSS9-ILQPDHks5thPuqgaJpZM4StOTh>
> .
>

Rich-Harris commented:
Released as 1.58.2. Thank you @jacwright!

ekhaled commented:
Can someone please post how to use this feature in full before it gets documented on the main website.

Whats the final syntax etc.

Conduitry commented:
@ekhaled The syntax is pretty well described in the initial PR description, but you can also take a look at the `action-*` tests in https://github.com/sveltejs/svelte/tree/master/test/runtime/samples

ekhaled commented:
Thanks @Conduitry.
As you can see many variations were discussed such as `with` and `enhancements` etc. So wasn't sure what was settled on finally.

jacwright commented:
@ekhaled yeah, I updated the description to the latest decision so people didn't have to wade through the comments. I should have made a note that it was updated, sorry about that!

jacwright commented:
Also, this PR should land soon too https://github.com/sveltejs/svelte/pull/1279 which adds some nuance to actions that ought to be included in the docs.

constgen commented:
So actions are separate modules or I should duplicate them on every component?

jacwright commented:
They can be separate modules. I hope we (the community) can build up some great actions in NPM that can be used by the community.
